### ISSUE TITLE
Selenium WebDriver driver.getSession() wait a long time

### ISSUE THREAD

**Issue by zixia** [Chatie/wechaty#86]
In our case, the timeout is `67` seconds: 
https://github.com/wechaty/wechaty/blob/41b29e702b5d1bbe4b35460845c44ef31ae416b8/src/puppet-web/browser-driver.ts#L254

Detail: https://travis-ci.org/wechaty/wechaty/jobs/175139957

I don't know why yet.

And it seems if we had an invalid WebDriver, then we will have a browser leak.

need to use `process.kill` to kill it, in `clean()`.

* Windows: https://github.com/nodejs/node-v0.x-archive/issues/8713#issuecomment-62769111

A unit test fail log like this:

```shell
VERB PuppetWebBrowser constructor() with head(chrome) sessionFile()
VERB PuppetWebBrowserDriver constructor(chrome)
VERB PuppetWebBrowserCookie constructor(BrowserDriver, )
VERB PuppetWebBrowser init()
VERB StateMonitor Browser:target(open) <- (close)
VERB StateMonitor Browser:current(open,false) <- (close,true)
VERB PuppetWebBrowserDriver init() for head: chrome
VERB PuppetWebBrowserDriver getChromeDriver()
VERB PuppetWebBrowserDriver getChromeDriver() new Builder()
VERB PuppetWebBrowserDriver getChromeDriver() new Builder() done
VERB PuppetWebBrowserDriver valid()
WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
WARN PuppetWebBrowserDriver valid() driver.getSession() exception: valid() driver.getSession() timeout(halt?)
VERB PuppetWebBrowserDriver getChromeDriver() valid() done: false
VERB PuppetWebBrowserDriver getChromeDriver() new Builder()
WARN PuppetWebBrowserDriver getChromeDriver() got invalid driver
WARN PuppetWebBrowserDriver getChromeDriver() with retry: 1
VERB PuppetWebBrowserDriver getChromeDriver() new Builder() done
VERB PuppetWebBrowserDriver valid()
VERB PuppetWebBrowserDriver valid() driver.getSession() done
VERB PuppetWebBrowserDriver valid() driver.executeScript() done
VERB PuppetWebBrowserDriver getChromeDriver() valid() done: true
VERB PuppetWebBrowser init() driver.init() done
VERB PuppetWebBrowser open(https://wx.qq.com/zh_CN/htmledition/v2/images/webwxgeticon.jpg)
VERB PuppetWebBrowser open(https://wx.qq.com)
VERB StateMonitor Browser:current(open,true) <- (open,false)
VERB PuppetWebBrowser open(https://wx.qq.com)
VERB PuppetWebBrowser quit()
VERB StateMonitor Browser:current(close,false) <- (open,true)
ERR PuppetWebBrowser retryPromise failed: clean() found browser process, not clean, dirty
WARN PuppetWebBrowser quit() exception: clean() found browser process, not clean, dirty
WARN PuppetWebBrowser driver.quit() exception: clean() found browser process, not clean, dirty
VERB StateMonitor Browser:current(close,true) <- (close,false)
  ✖ test › webdriver › WebDriver process create & quit test no driver process after quit   
  t.is(pids.length, 0, 'no driver process after quit')
            |                                         
            5                                         
  
```

**Comment by JasLin** [comment]
maybe cause by the latest version Selenium WebDriver ? it seems  time out take place more frequently after upgrade from 3.0.0-beta-1 to 3.0.0

**Comment by zixia** [comment]
@JasLin if this is true, please have a look at #72

**Comment by JasLin** [comment]
it's stranger , set SELENIUM_PROMISE_MANAGER=1 had as no effect on ```process.env.SELENIUM_PROMISE_MANAGER```


code : ```ding-dong-bot.ts```

```javascript
/**
 *
 * Wechaty - Wechat for Bot
 *
 * Connecting ChatBots
 * https://github.com/wechaty/wechaty
 *
 */

/* tslint:disable:variable-name */
const QrcodeTerminal = require('qrcode-terminal')

import {
  Wechaty
  , Config
  , log
} from '../'

const welcome = `
| __        __        _           _
| \\ \\      / /__  ___| |__   __ _| |_ _   _
|  \\ \\ /\\ / / _ \\/ __| '_ \\ / _\` | __| | | |
|   \\ V  V /  __/ (__| | | | (_| | |_| |_| |
|    \\_/\\_/ \\___|\\___|_| |_|\\__,_|\\__|\\__, |
|                                     |___/

=============== Powered by Wechaty ===============
-------- https://github.com/zixia/wechaty --------

I'm a bot, my super power is talk in Wechat.

If you send me a 'ding', I will reply you a 'dong'!
__________________________________________________

Hope you like it, and you are very welcome to
upgrade me for more super powers!

Please wait... I'm trying to login in...

`

console.log(welcome)

console.log('SELENIUM_PROMISE_MANAGER:' + process.env.SELENIUM_PROMISE_MANAGER)

const bot = Wechaty.instance({ profile: Config.DEFAULT_PROFILE })

bot
.on('login'	  , user => {
  log.info('Bot', `${user.name()} logined`)
  console.log('SELENIUM_PROMISE_MANAGER:' + process.env.SELENIUM_PROMISE_MANAGER)
})
.on('logout'	, user => log.info('Bot', `${user.name()} logouted`))
.on('error'   , e => log.info('Bot', 'error: %s', e))
.on('scan', (url, code) => {
  if (!/201|200/.test(String(code))) {
    let loginUrl = url.replace(/\/qrcode\//, '/l/')
    QrcodeTerminal.generate(loginUrl)
  }
  console.log(`${url}\n[${code}] Scan QR Code in above url to login: `)
})
.on('message', m => {
  try {
    const room = m.room()
    console.log((room ? '[' + room.topic() + ']' : '')
                + '<' + m.from().name() + '>'
                + ':' + m.toStringDigest()
    )

    if (/^(ding|ping|bing)$/i.test(m.content()) && !m.self()) {
      m.say('dong')
      log.info('Bot', 'REPLY: dong')
    }
  } catch (e) {
    log.error('Bot', 'on(message) exception: %s' , e)
  }
})

bot.init()
.catch(e => {
  log.error('Bot', 'init() fail: %s', e)
  bot.quit()
  process.exit(-1)
})

function logToFile(data) {
  require('fs').appendFile('message.log', data + '\n\n#############################\n\n', err => {
    if (err) { log.error('LogToFile: %s', err) }
  })
}
if (typeof logToFile === 'fasdfsd') {
  console.log('disable linting warning')
}

```

output:
```
   master ●   echo $SELENIUM_PROMISE_MANAGER
0
    master ●  export SELENIUM_PROMISE_MANAGER=1 && ts-node example/ding-dong-bot.ts

| __        __        _           _
| \ \      / /__  ___| |__   __ _| |_ _   _
|  \ \ /\ / / _ \/ __| '_ \ / _` | __| | | |
|   \ V  V /  __/ (__| | | | (_| | |_| |_| |
|    \_/\_/ \___|\___|_| |_|\__,_|\__|\__, |
|                                     |___/

=============== Powered by Wechaty ===============
-------- https://github.com/zixia/wechaty --------

I'm a bot, my super power is talk in Wechat.

If you send me a 'ding', I will reply you a 'dong'!
__________________________________________________

Hope you like it, and you are very welcome to
upgrade me for more super powers!

Please wait... I'm trying to login in...


SELENIUM_PROMISE_MANAGER:0
INFO Wechaty v#git[f38fac0 code clean] initializing...
WARN PuppetWebBrowserCookie addCookies() exception: invalid cookie domain: invalid domain:".wx2.qq.com"
  (Session info: chrome=54.0.2840.90)
  (Driver info: chromedriver=2.24.417424 (c5c5ea873213ee72e3d0929b47482681555340c3),platform=Linux 4.4.0-45-generic x86_64)
ERR PuppetWebBrowserCookie load() add() exception: invalid cookie domain: invalid domain:".wx2.qq.com"
  (Session info: chrome=54.0.2840.90)
  (Driver info: chromedriver=2.24.417424 (c5c5ea873213ee72e3d0929b47482681555340c3),platform=Linux 4.4.0-45-generic x86_64)
^C
     master ●  echo $SELENIUM_PROMISE_MANAGER
1
```

**Comment by JasLin** [comment]
found it hardcode in ```config.ts```

```
process.env['SELENIUM_PROMISE_MANAGER'] = 0
```

will try to change it in code

**Comment by JasLin** [comment]
changed it in code . 
can not see the difference ,both are slow.  i will trying to  change the code to calculate the time spent on getSession() when i  am back afternoon.

**Comment by JasLin** [comment]
@zixia 

### code

```javascript

 private async getChromeDriver(): Promise<WebDriver> {
    log.verbose('PuppetWebBrowserDriver', 'getChromeDriver()')
       ......
        console.time('valid')
        const valid = await this.valid(driver)
        console.timeEnd('valid')
        ......
    return driver
 }

private async valid(driver: WebDriver): Promise<boolean> {
    log.verbose('PuppetWebBrowserDriver', 'valid()')
    try {
      ......
        console.time('getSession')        
        // resolve
        driver.getSession()
              .then(session => {
                clearTimeout(timer)
                resolve(session)
              })
        console.timeEnd('getSession')
      })
     ......
  }
```

### output



#### SELENIUM_PROMISE_MANAGER:0


```
getSession: 537.926ms
valid: 546.636ms

getSession: 491.898ms
valid: 500.848ms

getSession: 511.089ms
valid: 519.051ms

getSession: 487.795ms
valid: 500.454ms

getSession: 498.500ms
valid: 509.621ms

```

#### SELENIUM_PROMISE_MANAGER:1

```
getSession: 490.944ms
valid: 501.270ms

getSession: 484.682ms
valid: 496.679ms

getSession: 497.962ms
valid: 507.867ms

getSession: 504.996ms
valid: 514.519ms
```

it looks everything all right.

**Comment by JasLin** [comment]
@zixia another problem, on starting robot. opening the ```qrcode url``` at the chrome driver  is more slow then opening at  the native chrome browser. 

in my case ,the native chrome popup and  qrcode  show fast(immediately), but the qrcode show in the terminator  was show after about 26s.

time trace: in ```broswer.ts``` 

## code

```javascript
public async init(): Promise<void> {
    log.verbose('PuppetWebBrowser', 'init()')
    ......
    try {
      await this.driver.init()
      log.verbose('PuppetWebBrowser', 'init() driver.init() done')
      console.time('browser.openJumpUrl')
      await this.open(jumpUrl)
      console.timeEnd('browser.openJumpUrl') 
      await this.loadCookie()
                .catch(e => { // fail safe
                  log.verbose('PuppetWebBrowser', 'browser.loadSession(%s) exception: %s'
                                          , this.setting.sessionFile
                                          , e && e.message || e
                  )
                })
      console.time('browser.open')  
      await this.open()
      console.timeEnd('browser.open')      
      ......
    }
  }

 public async open(url: string = 'https://wx.qq.com'): Promise<void> {
    log.verbose('PuppetWebBrowser', `open(${url})`)
    try {
      console.time('driver.get')
      await this.driver.get(url)
      console.timeEnd('driver.get')
    } catch (e) {
      log.error('PuppetWebBrowser', 'open() exception: %s', e.message)
      throw e
    }
  }
```

## output (ran it several times )

```
driver.get: 270.763ms
browser.openJumpUrl: 272.420ms
driver.get: 26375.641ms
browser.open: 26376.105ms
```

```
driver.get: 277.540ms
browser.openJumpUrl: 278.158ms
driver.get: 26444.618ms
browser.open: 26445.229ms
```

**only this time is fast**

```
driver.get: 623.712ms
browser.openJumpUrl: 624.237ms
driver.get: 2792.972ms
browser.open: 2793.391ms
```

```
driver.get: 291.821ms
browser.openJumpUrl: 292.312ms
driver.get: 26754.434ms
browser.open: 26754.965ms
```

```
driver.get: 264.332ms
browser.openJumpUrl: 264.723ms
driver.get: 29616.550ms
browser.open: 29617.324ms
```

**Comment by zixia** [comment]
Thanks for the benchmark.

About the `26s` problem, it is a known issue #2 , I believe it belongs to the WebDriver support of browser. maybe need to hack `chromedriver` or `ghostdriver`(phantomjs) to know how to skip the waiting code.

**Comment by zixia** [comment]
Still experiencing this issue recently.

https://travis-ci.org/Chatie/wechaty/jobs/237692304#L1626
https://travis-ci.org/Chatie/wechaty/jobs/237692304#L1697

**Comment by zixia** [comment]
This bug is still paining me time and again.

**Comment by zixia** [comment]
This issue should be fixed at v0.10 or above.
